src/schedule/ucc_schedule.c

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+#include "ucc_schedule.h"
+#include "utils/ucc_compiler_def.h"
+
+void ucc_event_manager_init(ucc_event_manager_t *em)
+{
+    int i;
+    for (i = 0; i < UCC_EVENT_LAST; i++) {
+        em->listeners_size[i] = 0;
+    }
+}
+
+void ucc_event_manager_subscribe(ucc_event_manager_t *em, ucc_event_t event,
```

Comment:
return ucc_status_t

Reply:
again, don't need. i'll add ucc-assert to check for MAX_LISTENERS. Other than that there is nothing to fail here. We will implement algorithms internally. So we now we should not overflow em.

I put some thought into it. On the one hand we can implement generic support for arbitrary number of listeners. We can put some preallocated number of listeners into em (like it is right now) and add parameter "num_listeners" to coll_task_init. Then if task knows it needs more listeners there will be dynamic memory allocation to hold them (malloc/free). Obviously we will also need a ucc_coll_task_cleanup fn to release it. Now the question is: if we do that, will we have any algorithms that really need it ? I mean, it is better writing the algorithm with less dependencies between tasks (not exceeding max number of listeners) rather than doing malloc/free in a critical path.